src/scheduler/gc_work.rs

```diff
@@ -815,6 +920,27 @@ impl<VM: VMBinding, P: PlanTraceObject<VM> + Plan<VM = VM>, const KIND: TraceKin
             slot.store(new_object);
         }
     }
+
+    /// Start the a scan work packet. If SCAN_OBJECTS_IMMEDIATELY, the work packet will be executed immediately, in this method.
+    /// Otherwise, the work packet will be added the Closure/ClosureImmovable work bucket and will be dispatched later by the scheduler.
+    #[inline]
+    fn start_or_dispatch_scan_work(&mut self, work_packet: impl GCWork<Self::VM>) {
+        if Self::SCAN_OBJECTS_IMMEDIATELY {
+            // We execute this `scan_objects_work` immediately.
+            // This is expected to be a useful optimization because,
+            // say for _pmd_ with 200M heap, we're likely to have 50000~60000 `ScanObjects` work packets
+            // being dispatched (similar amount to `ProcessEdgesWork`).
+            // Executing these work packets now can remarkably reduce the global synchronization time.
+            self.worker().do_work(work_packet);
+        } else {
+            if KIND == TRACE_KIND_IMMOVABLE {
```

Comment:
Similar to `ObjectsClosure`, you can just have a field to store which bucket new work should go to.